src/main/java/io/quarkus/gizmo2/impl/EqualsHashCodeToStringGenerator.java

```diff
@@ -0,0 +1,146 @@
+                b0.if_(b0.eq(mc.this_(), other), BlockCreator::returnTrue);
+                b0.ifNotInstanceOf(other, thisClass, BlockCreator::returnFalse);
+
+                Expr otherCast = b0.define("other", b0.cast(other, thisClass));
+                for (FieldDesc field : fields) {
+                    if (!field.owner().equals(thisClass)) {
+                        throw new IllegalArgumentException("Field does not belong to " + thisClass.displayName() + ": " + field);
+                    }
+
+                    LocalVar thisValue = b0.define("thisValue", b0.get(mc.this_().field(field)));
+                    LocalVar thatValue = b0.define("thatValue", b0.get(otherCast.field(field)));
+                    String fieldDesc = field.type().descriptorString();
+                    switch (fieldDesc.charAt(0)) {
+                        // boolean, byte, short, int, long, char
+                        case 'Z', 'B', 'S', 'I', 'J', 'C' -> b0.if_(b0.ne(thisValue, thatValue), BlockCreator::returnFalse);
+                        // float
+                        case 'F' -> {
+                            // this is consistent with Arrays.equals() and it's also what IntelliJ generates
+                            Expr thisBits = b0.invokeStatic(floatToIntBits, thisValue);
+                            Expr thatBits = b0.invokeStatic(floatToIntBits, thatValue);
+                            b0.if_(b0.ne(thisBits, thatBits), BlockCreator::returnFalse);
+                        }
+                        // double
+                        case 'D' -> {
+                            // this is consistent with Arrays.equals() and it's also what IntelliJ generates
+                            Expr thisBits = b0.invokeStatic(doubleToLongBits, thisValue);
+                            Expr thatBits = b0.invokeStatic(doubleToLongBits, thatValue);
+                            b0.if_(b0.ne(thisBits, thatBits), BlockCreator::returnFalse);
+                        }
+                        // Object
+                        case 'L' -> b0.unless(b0.exprEquals(thisValue, thatValue), BlockCreator::returnFalse);
+                        // array
+                        case '[' -> b0.unless(b0.arrayEquals(thisValue, thatValue), BlockCreator::returnFalse);
+                        default -> throw impossibleSwitchCase(field);
+                    }
+                }
+
+                b0.return_(true);
+            });
+        });
+    }
+
+    public void generateHashCode() {
+        ClassDesc thisClass = cc.type();
+
+        cc.method("hashCode", mc -> {
+            mc.public_();
+            mc.returning(int.class);
+            mc.body(b0 -> {
+                if (fields.isEmpty()) {
+                    b0.return_(0);
+                    return;
+                }
+
+                LocalVar result = b0.define("result", Constant.of(1));
+                for (FieldDesc field : fields) {
+                    if (!field.owner().equals(thisClass)) {
```

Comment:
You could conceivably include fields of a superclass, right?

Reply:
I didn't think about that, but you're right. I'm not really keen on removing the check, but I don't see a better way...

Reply:
So I'm thinking handling superclasses is out of scope, at least for now. The best way to handle superclasses would be to call `equals()` (or `hashCode()`; not sure about `toString()`) on the superclass first, and then handle the fields of this class.